update to 'Fix google map offsets in china'

johnd0e opened the pull request:
Misc updates for #68.

fix-china-map-offset: Baidu support

* Sources refactored to be closer to upstream https://github.com/Artoria2e5/PRCoords
* ~Baidu support~
* Clarify comment about usage

...and more:
* applied to googleMutant
* fix filename

Commit:
fix-china-map-offset: Baidu support

* Sources refactored to be closer to upstream https://github.com/Artoria2e5/PRCoords
* Baidu support
* Clarify comment about usage

Commit:
fix-china-map-offset: appied to googleMutant

Commit:
fix-china-map-offset: fix filename

Commit:
fix-china-map-offset: fixme: options for HongKong and map type

Commit:
fixup! fix-china-map-offset: fixme: options for HongKong and map type

Commit:
fix-china-map-offset: refactor (intermediate function eliminated)

Commit:
fix-china-map-offset: refactor (redefine leaflet methods cleaner)

johnd0e commented:
Todo:
1. > applied to googleMutant

   Perhaps it wasn't necessary and should be reverted.

2. > options for HongKong...

   It seems that it needed for Baidu only.
   If we could make plugin for Baidu it would have sense.
   But I'm not sure.
   (If not than this could be removed)

3. > ...and map type

   We need to test different maps.
   Perhaps this required for Google only, so condition need to be enhanced.

4. It also possible that we do not need check _'if in China'_ for every Chinese map.
   Need more maps' plugins to test...
   BTW, leaflet has interesting [function](http://leafletjs.com/reference.html#lineutil-simplify), that allows to simplificate our polygon, like here: http://mourner.github.io/simplify-js/ 

5. > (redefine leaflet methods cleaner)

   I suppose that we can make it even cleaner, not touching `L.GridLayer.prototype`.
  Will try tomorrow.

modos189 commented on code/boot.js:
```diff
@@ -141,7 +141,9 @@ function createDefaultBaseMapLayers() {
+        ],
+      needFixChinaOffset: true // managed by fix-china-map-offset plugin
+    });
   baseLayers['Google Roads'] = L.gridLayer.googleMutant({type:'roadmap', maxZoom: 21});
```
Here and below, too need add
`needFixChinaOffset: true`

johnd0e replied:
Right.
So, we can add this option here, only for that layers that need it (excl. `hybrid` and `satellite`).
(But I am also going to try another way)

Commit:
fix-china-map-offset: cosmetics (restore original source indents)

Commit:
fix-china-map-offset: change leaflet extending

do not touch L.GridLayer
extend L.TileLayer and L.GridLayer.GoogleMutant

Commit:
fix-china-map-offset: set 'needFixChinaOffset' option in boot.js

johnd0e commented:
I think that better leave boot.js untouched and patch layers options in plugin setup instead.
All layers are accessible via layerChooser._layers, and it's easy to patch only _mutant layers.
Will try tomorrow.

Commit:
fix-china-map-offset: fix: put actions into setup function

Commit:
fix-china-map-offset: patch GoogleMutant in plugin completely (not in boot.js)

johnd0e commented:
@modos189 
I need an advice.

Currently we rewrite `L.TileLayer` class:
```
  L.TileLayer.include(fixChinaOffset);
```

A more proper way would be to extent it with new class:
```
  L.TileLayer.GCG02 = L.TileLayer.extend(fixChinaOffset);
```

And then instead of `needFixChinaOffset` option we would use that new class:
```
var mapLayer = new L.TileLayer.GCG02 (url, options);
```

But I doubt that we should make things more complex here..

Commit:
fixup! fix-china-map-offset: cosmetics (restore original source indents)

Commit:
fix-china-map-offset: remove Baidu-stuff

In current state it doesn't function.
Maybe someone reimplement it in future.
More info here: #64.

johnd0e commented:
@modos189?

Commit:
fix-china-map-offset: shorten 'namespace' access

Commit:
fix-china-map-offset: fix: L.Util may not be available yet

Commit:
fix-china-map-offset: fix: namespace conflict
